src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -183,6 +122,153 @@ public GeneratorDriver AddAdditionalTexts(ImmutableArray<AdditionalText> additio
+
+            static ImmutableArray<GeneratedSourceResult> getSourceResults(GeneratorState generatorState)
+            {
+                ArrayBuilder<GeneratedSourceResult> sourceResults = ArrayBuilder<GeneratedSourceResult>.GetInstance();
+                for (int i = 0; i < generatorState.SourceTexts.Length; i++)
+                {
+                    var sourceText = generatorState.SourceTexts[i];
+                    var tree = generatorState.Trees[i];
+
+                    sourceResults.Add(new GeneratedSourceResult(tree, sourceText.Text, sourceText.HintName));
+                }
+                return sourceResults.ToImmutableAndFree();
+            }
+        }
+
+        internal GeneratorDriverState RunGeneratorsCore(Compilation compilation, DiagnosticBag diagnosticsBag, CancellationToken cancellationToken = default)
+        {
+            // with no generators, there is no work to do
+            if (_state.Generators.IsEmpty)
+            {
+                return _state;
+            }
+
+            // run the actual generation
+            var state = StateWithPendingEditsApplied(_state);
+            var stateBuilder = ArrayBuilder<GeneratorState>.GetInstance();
+            int receiverCount = 0;
+
+            for (int i = 0; i < state.Generators.Length; i++)
+            {
+                var generator = state.Generators[i];
+                var generatorState = state.GeneratorStates[i];
+
+                // initialize the generator if needed
+                if (!generatorState.Info.Initialized)
+                {
+                    InitializationContext context = new InitializationContext(cancellationToken);
+                    Exception? ex = null;
+                    try
+                    {
+                        generator.Initialize(context);
+                    }
+                    catch (Exception e)
+                    {
+                        ex = e;
+                    }
+                    generatorState = ex is null
+                                     ? new GeneratorState(context.InfoBuilder.ToImmutable())
+                                     : SetGeneratorException(new GeneratorState(), generator, ex, diagnosticsBag, isInit: true);
+                }
+
+                // create the syntax receiver if requested
+                if (generatorState.Info.SyntaxReceiverCreator is object)
+                {
+                    ISyntaxReceiver? rx = null;
+                    try
+                    {
+                        rx = generatorState.Info.SyntaxReceiverCreator();
+                    }
+                    catch (Exception e)
+                    {
+                        generatorState = SetGeneratorException(generatorState, generator, e, diagnosticsBag);
+                    }
+
+                    if (rx is object)
+                    {
+                        generatorState = generatorState.WithReceiver(rx);
+                        receiverCount++;
+                    }
```

Comment:
Consider moving this inside the `try` body to make it clearer that it only has impact in non-exception cases.

Reply:
:p chuck suggested I move it out. Reason being that if there's an exception caused by *our* code we'll report it as being the generators fault.